lib/spack/spack/operating_systems/windows_os.py

```diff
@@ -83,11 +85,25 @@ def compiler_search_paths(self):
                     os.path.join(str(os.getenv("ONEAPI_ROOT")), "compiler", "*", "windows", "bin")
                 )
             )
-        # Second strategy: Find MSVC via the registry
-        msft = winreg.WindowsRegistryView(
-            "SOFTWARE\\WOW6432Node\\Microsoft", winreg.HKEY.HKEY_LOCAL_MACHINE
-        )
-        vs_entries = msft.find_subkeys(r"VisualStudio_.*")
+        tries = 0
+        while tries < 3:
+            try:
+                # Second strategy: Find MSVC via the registry
+                # Registry interactions are subject to race conditions, etc and can generally
+                # be flakey, do this in a catch block to prevent reg issues from interfering
+                # with compiler detection
+                msft = winreg.WindowsRegistryView(
+                    "SOFTWARE\\WOW6432Node\\Microsoft", winreg.HKEY.HKEY_LOCAL_MACHINE
+                )
+                vs_entries = msft.find_subkeys(r"VisualStudio_.*", depth=False)
+                break
+            except OSError as e:
+                if tries == 2:
+                    tty.debug(
+                        'Windows registry query on "SOFTWARE\\WOW6432Node\\Microsoft"'
+                        f"under HKEY_LOCAL_MACHINE: {str(e)}"
+                    )
+                tries += 1
```

Comment:
A couple questions:

* Is this likely to succeed if re-executed immediately? Should this sleep for a short while?
* How do programs generally handle this periodic unavailability?

Reply:
So on further investigation, the slate of unexpected errors (those outside of `FileNotFound` and `InvalidPermissions`) can be raised for a couple of reasons, some of which are race conditions, but others of which are (AFAIU, I read a couple forum threads) effectively corner cases of permission issues where some programs can write Keys that are "readable" for your user given the ACL for the entry, but are actually locked.

Given that in some cases this is never going to go away (we just can't read the key), then I think we absolutely don't want to sleep here. As I understand it, the race conditions are generally caused by some monitor software (anti-malware or otherwise) reading the keys and should be resolved more or less instantaneously. 

All of the python reg interactions I was able to find on GH implement more or less the same approach we do here, `try/except` blocks.

With all that in mind, I think we should not sleep, and only retry once. Thoughts?